### PR TITLE
[PotentialFlow] Adding support for MLMC with XMC in stochastic optimization

### PR DESCRIPTION
**Pull request by marcnunezc**
**Description**
This does the necessary changes to use the MLMC algorithm with bodyfitted meshes in optimization, which uses different levels of refinement to perform the UQ analysis. This requires the mapping app to send the data across levels to a common discretization.

Also an option to save to file the statisical error, runtime, etc is added. 

Please mark the PR with appropriate tags: 
- Api Breaker, Mpi, etc...

**Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- Adding support for MLMC with xmc
- Adding option to store statistical data to a dict
- Added a nightly test

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py [hunk 1]
```diff
@@ -3,13 +3,14 @@
...
 import KratosMultiphysics.CompressiblePotentialFlowApplication.potential_flow_analysis as potential_flow_analysis
+import KratosMultiphysics.MappingApplication
 
 # Import Kratos, XMC, PyCOMPSs API
 import KratosMultiphysics.MultilevelMonteCarloApplication
 import xmc
 import xmc.methodDefs_momentEstimator.computeCentralMoments as mdccm
 from exaqute import get_value_from_remote
-import json, os
+import json, os, math, time
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py]: I suggest to first import Python libraries and then import Kratos/xmc libraries. @roigcarlo taught me so time ago.

File: applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py [hunk 2]
```diff
@@ -99,23 +111,41 @@ def InitializeSolutionStep(self):
...
             estimator_container.append(float(get_value_from_remote(self.xmc_analysis.monteCarloSampler.indices[index].qoiEstimator[qoi_counter].value(order=order, isCentral=is_central))))
+            error_container.append(float(get_value_from_remote(self.xmc_analysis.monteCarloSampler.indices[index].qoiEstimator[qoi_counter].value(order=order, isCentral=is_central, isErrorEstimationRequested=True)[1])))
+            n_samples_container.append(int(get_value_from_remote(self.xmc_analysis.monteCarloSampler.indices[index].qoiEstimator[qoi_counter]._sampleCounter)))
         qoi_counter += 1
         # linearly sum estimators: this summation operation is valid for expected value and central moments
         # we refer to equation 4 of Krumscheid, S., Nobile, F., & Pisaroni, M. (2020). Quantifying uncertain system outputs via the multilevel Monte Carlo method — Part I: Central moment estimation. Journal of Computational Physics. https://doi.org/10.1016/j.jcp.2020.109466
         self._value = sum(estimator_container)
+        statistical_error = math.sqrt(sum(error_container))
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py:142]:
```suggestion
        # compute statistical error as in section 2.2 of
        # Pisaroni, M., Nobile, F., & Leyland, P. (2017). A Continuation Multi Level Monte Carlo (C-MLMC) method for uncertainty quantification in compressible inviscid aerodynamics. Computer Methods in Applied Mechanics and Engineering, 326, 20–50. https://doi.org/10.1016/j.cma.2017.07.030
        statistical_error = math.sqrt(sum(error_container))
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py [hunk 3]
```diff
@@ -99,23 +111,41 @@ def InitializeSolutionStep(self):
...
             order = 1 ; is_central = False
         elif self.risk_measure == "variance":
             order = 2 ; is_central = True
 
+        if not self.output_dict_results_file_name == "":
+            self.results_dict[self.step] = {}
+
         # save lift coefficient
         qoi_counter = 0
         estimator_container = [] # here we append contribution for each index/level
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py]:
```suggestion
        estimator_container = [] # here we append the estimator for each index/level
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py [hunk 4]
```diff
@@ -99,23 +111,41 @@ def InitializeSolutionStep(self):
...
+        error_container = [] # here we append contribution for each index/level
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py]:
```suggestion
        error_container = [] # here we append the variance of the estimator for each index/level
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py [hunk 5]
```diff
@@ -263,6 +295,9 @@ def __init__(self,input_model,input_parameters,sample):
         self.mapping = False
         self.adjoint_parameters_path =input_parameters["adjoint_parameters_path"].GetString()
         self.design_surface_sub_model_part_name = input_parameters["design_surface_sub_model_part_name"].GetString()
+        self.main_model_part_name = input_parameters["solver_settings"]["model_part_name"].GetString()
+        self.auxiliary_mdpa_path = input_parameters["auxiliary_mdpa_path"].GetString()
+
         super(SimulationScenario,self).__init__(input_model,input_parameters)
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py]:
```suggestion
        super().__init__(input_model,input_parameters)
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py [hunk 6]
```diff
@@ -280,27 +322,27 @@ def Finalize(self):
...
         """
         Method introducing the stochasticity in the right hand side. Mach number and angle of attack are random varaibles.
         """
-        mach = self.sample[1]
-        alpha =  self.sample[2]
+        mach = abs(self.sample[1])
+        alpha = self.sample[2]
         self.project_parameters["processes"]["boundary_conditions_process_list"][0]["Parameters"]["mach_infinity"].SetDouble(mach)
         self.project_parameters["processes"]["boundary_conditions_process_list"][0]["Parameters"]["angle_of_attack"].SetDouble(alpha)
         super(SimulationScenario,self).ModifyInitialProperties()
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py]:
```suggestion
        super().ModifyInitialProperties()
```

File: applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/refinement_parameters.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/refinement_parameters.json:1]: Indentation can be improved.

File: applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_parameters.json [hunk 8]
```diff
@@ -53,7 +53,7 @@
     "monoCriteriaInpuctDictionary" :{
         "statisticalError": {
             "criteria": "xmc.methodDefs_monoCriterion.criterionFunctions.isLowerThanOrEqualTo",
-            "tolerance": [1.1e-03],
+            "tolerance": [100.0],
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_parameters.json]: Why this change? If you want to be sure that you do not do more than one iteration, you can change the key `maxNumberIterations`.

**Review by riccardotosi — COMMENTED** [review]
Some comments. Have you run manually the nightly CI?

File: applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_parameters.json (hunk 8 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_parameters.json]: Ok thanks!

**Comment by marcnunezc** [comment]
> Some comments. Have you run manually the nightly CI?

Yes, it is running here:

https://github.com/KratosMultiphysics/Kratos/actions/runs/981973413

File: applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        "outputBatchSize": 1,
+        "projectParametersPath": "primal_parameters.json",
+        "refinementParametersPath": "refinement_parameters.json",
+        "refinementStrategy": "stochastic_adaptive_refinement",
+        "sizeMultiXMomentEstimator": 12
+    },
+    "monoCriteriaInpuctDictionary" :{
+        "statisticalError": {
+            "criteria": "xmc.methodDefs_monoCriterion.criterionFunctions.isLowerThanOrEqualTo",
+            "tolerance": [100.0],
```
Comment (riccardotosi) [inline applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json:67]: Same.

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json:67]: Here the max iterations are correclty set already: 

https://github.com/KratosMultiphysics/Kratos/blob/c2e956975a2f263e50dc774c571765cbfd43228a/applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json#L77

Comment (riccardotosi) [reply applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json:67]: I am referring to the tolerance, that is very high. I would place here the tolerance that you are actually satisfying.

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json:67]: I dont really see the point, the tolerance is kinda meaningless here, the main interest is just to go through the code once for testing purposes (which is anyway controlled by the maxIter key). I changed it back in the other json to reduce the changes since the other value was already in place.

Comment (riccardotosi) [reply applications/CompressiblePotentialFlowApplication/tests/stochastic_body_fitted_opt/xmc_mlmc_parameters.json:67]: As you prefer, I was commenting only because if an external user sees this he/she may wonder why it is so high.

**Review by riccardotosi — DISMISSED** [review]
Ok if nightly passes.

**Comment by riccardotosi** [comment]
> > Some comments. Have you run manually the nightly CI?
> 
> Yes, it is running here:
> 
> https://github.com/KratosMultiphysics/Kratos/actions/runs/981973413

~~~bash
 Running CompressiblePotentialFlowApplication tests
......s..sss/__w/Kratos/Kratos/bin/Release/applications//CompressiblePotentialFlowApplication/tests/test_CompressiblePotentialFlowApplication.py
Completed CompressiblePotentialFlowApplication tests with exit code 1!
~~~

**Comment by riccardotosi** [comment]
Just let me know when nightly passes and I can re-approve.

**Comment by marcnunezc** [comment]
This should be ready to merge! Nighly worked fine (ignoring CLawApp) here:

https://github.com/KratosMultiphysics/Kratos/runs/3021989672?check_suite_focus=true

**Review by riccardotosi — APPROVED** [review]
